src/RollbarClient.test.mts

```diff
@@ -1,23 +1,28 @@
+// External Imports
+import { beforeAll, describe, expect, test, vi } from 'vitest';
+
 // Internal Imports
-import { RollbarClient } from './RollbarClient.js';
+import { RollbarClient } from './RollbarClient.mjs';
+import type { IConfigurationOptions } from './interfaces.mjs';
+import type { TSubmitterParameters } from './types.mjs';
 
 // Module Mocks
-const mockReport = jest.fn();
-jest.mock('./RollbarClientSubmitter.js', () => ({
-  RollbarClientSubmitter: jest.fn().mockImplementation(() => ({
-    report: mockReport,
-  })),
-}));
+vi.mock('./RollbarClientSubmitter', () => {
+  const RollbarClientSubmitter = vi.fn();
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access -- TODO: determine how to type prototype as non-any
```

Comment:
I was stuck for a long time on this where it was complaining about `.prototype` being `any`; I eventually decided to move on. I don't think this is a huge deal, but if you have any ideas I'll certainly consider them.

Reply:
I thought so too, but the Vitest documentation specifically advises to uses prototype in this way to mock classes (see "Mock exported class implementation"): https://vitest.dev/guide/mocking.html#:~:text=Mock%20exported%20class%20implementation

Using `class` causes the same error too confusingly

Reply:
Huh interesting.

In this same docs block they also have:

> Example with vi.mock and return value:

> ts
import { SomeClass } from 'some-path'
vi.mock('some-path', () => {
  const SomeClass = vi.fn(() => ({
    someMethod: vi.fn()
  }))
  return { SomeClass }
})
// SomeClass.mock.returns will have returned object

So I guess we could try that instead.

Though if the class gives the same error as the prototype, I think it means the problem is not with `prototype` but with the type of `RollbarClientSubmitter` within this snippet, which would make sense as it is never defined what type `RollbarClientSubmitter` should be here.

So I think you probably need something liket his if `vi.fn` is a generic:
`const RollbarClientSubmitter = vi.fn<RollbarClientSubmitter>();` (where the `RollbarClientSubmitter` between chevrons should be the original one as we want to use it as a type)
or if `vi.fn` isn't a generic:
`const RollbarClientSubmitter = vi.fn() as RollbarClientSubmitter;`

I think that should rid the code of the warning.

Reply:
Ugh classes just suck 😅 

After playing a little bit with it, the best I can do is:
```typescript
  const MockRollbarClientSubmitter = vi.fn(() => ({
    report: vi.fn(),
  } as Partial<RollbarClientSubmitter>));
  return { RollbarClientSubmitter: MockRollbarClientSubmitter };
```

NB: I need the `Partial` as the mock doesn't implement all the functions exposed by `RollbarClientSubmitter`, while it's ok, it's also risky as technically nothing would prevent `RollbarClient` to use more things from `RollbarClientSubmitter` as they aren't private.
I think, if internals were marked as private, then the Partial could be dropped, but somehow you need to extract the keys instead of using the type directly like so:

```typescript
  const MockRollbarClientSubmitter = vi.fn(() => ({
    report: vi.fn(),
  } as {[K in keyof RollbarClientSubmitter]: RollbarClientSubmitter[K]}));
  return { RollbarClientSubmitter: MockRollbarClientSubmitter };
 ```

Reply:
Haha. Classes can be confounding for sure.

Thank you for giving this a shot! My concern here is you have not mimic'd a class here, rather when instantiating a new object a new `report` function object will be created whereas attaching it to `prototype` will reuse the same function object for every instantiation. I'd like to stick to a class-like data structure here as closely as possible. What are your thoughts on this?

Good suggestion about private class fields, but unfortunately Safari support isn't quite far enough along relative to our userbase for me to feel comfortable using it yet: https://caniuse.com/mdn-javascript_classes_private_class_fields

If I were to use private class fields, I would have to move some of my tests into the code files themselves a.k.a. "in-source testing" (https://vitest.dev/guide/in-source.html, Rust also does this which is where they got the idea I think) which I need to think more about. Basically anything inaccessible from an external module scope would have to be tested that way. This is something Jest does not provide that I plan to play with in the future.

Reply:
> Thank you for giving this a shot! My concern here is you have not mimic'd a class here, rather when instantiating a new object a new report function object will be created whereas attaching it to prototype will reuse the same function object for every instantiation. I'd like to stick to a class-like data structure here as closely as possible. What are your thoughts on this?

You could get one function handling all also with the structure above
```typescript
const MockRollbarClientSubmitterPrototype = {
  report: vi.fn(),
};
const MockRollbarClientSubmitter = vi.fn(() => MockRollbarClientSubmitterPrototype);
return { RollbarClientSubmitter: MockRollbarClientSubmitter };
```
but yeah there are probably other differences between class VS object, though chances are you also already have some differences in the current code (because `vi.fn()` isn't a class).

For private class fields, you could use them in TypeScript but transpile them (that's what I had in mind personally)

But just in case none of this is a blocker for me, ultimately we are typing things in a test, it can be nice to get the types right as it'll make writing tests easier, but it's clearly not the most important.

Reply:
Thank you for your useful feedback here, but I'd like to leave this as-is as a reminder to iterate on this a bit in the future. I will keep your comments in mind for sure.

For private class fields, because I'm not using `tsc` to build the code, I can't specify compatibility targets. I'd have to look to see what `esbuild`'s built-in transpilation target is. I know it's configurable, but it is not compatible with `browserslist` which is something I'd like to examine more in the future. I'll keep `#private_field`s in mind (I assume you were not talking about `private field` which is not a runtime private field).

Reply:
I was thinking of `private field`, while I agree it's useless for JavaScript, it isn't useless for TypeScript ;)

Reply:
`#private` gives you the benefits in both places which was why I was interested in using it